File: C/Trees/BinaryTreeToBST.c
#include <stdio.h>
#include <stdlib.h>
 
 
/*
Approach:
Using array in C to store and sort values of given Binary Tree and then iterating through that array to get the BST.
All traversal in the following code has been done inorder.
*/

 
/* Structure for the Node */
struct node
{
    int info;
    struct node* left, *right;
};
 
int data[100];
int i = 0;
 
/*creates a new node */
 
struct node* genNode(int key)
{
    struct node* root = (struct node*)malloc(sizeof(struct node));
    root->info = key;
    root->left = NULL;
    root->right = NULL;
 
    return(root);
}
 
/* counts the number of nodes in a tree */
 
int count(struct node *n)
{
    int c = 1;
 
    if (n == NULL)
        return 0;
    else
    {
        c += count(n->left);
        c += count(n->right);
        return c;
    }
}
 
/*
 * Copies the nodes of Binary Tree in an array
 */
void binarytoarray(struct node * root, int data[], int *ptr)
{
    if(root != NULL)
    {
    binarytoarray(root->left, data, ptr);
    data[*ptr] = root->info;
    (*ptr)++;
    binarytoarray(root->right, data, ptr);
    }
}
 
/*
 * Copies the element of array to bst
 */
void arraytobst(int *arr, struct node *root, int *index)
{
    if(root != NULL)
    {
        arraytobst(arr, root->left, index);
        root->info = arr[i++];  // This i is a global variable initialized in line 20
        arraytobst(arr, root->right, index);
    }
}
 
int compare(const void *a, const void *b)
{
    return *(int*)a-*(int*)b;
}
 
/*
 * Function to traverse the tree (Inorder)
 */
 
void traverse(struct node *root)
{
    if(root !=NULL)
    {
        traverse(root->left);
        printf("%d\t",root->info);
        traverse(root->right);
    }
}
 
/*
 * Function to converts binary tree to bst
 */
void binaryToBST(struct node *root)
{
    int n, i;
 
    if (root == NULL)
        return;
    n = count(root);
    i = 0;
    binarytoarray(root, data, &i);
    qsort(&data, n, sizeof(data[0]), compare);
    i = 0;
    arraytobst(data, root, &i);
}
 

int main()
{
    struct node *root = genNode(9);
    root->left = genNode(4);
    root->right = genNode(5);
    root->left->left = genNode(15);
    root->right->right= genNode(1);
    root->right->left = genNode(3);
    root->right->right = genNode(7);
 
 
    printf("Inorder Traversal of Input Binary Tree is\n");
    traverse(root);
    binaryToBST(root);
    printf("\nInorder Traversal of the converted Binary Search Tree is\n");
    traverse(root);
    return 0;
}

File: Java/Trees/BinaryTreeToBST.java
import java.util.*;

/* Approach:
Using TreeSet in java to sort and store values of given Binary Tree and then iterating through that set to get the BST.
All traversal in the following code has been done inorder.
*/

class BinaryTreeToBST {
  // Class for defining structure of Nodes of Tree
  public class Node {
    int data;
    Node left, right;

    Node(int data) {
      this.data = data; // Initializing the current node data
      this.left = null; // Making left child null
      this.right = null; // Making right child null
    }
  }
  // Function to perform traversal of Tree (Inorder)
  public static void traverse(Node root) {
    if (root == null) {
      return;
    }

    traverse(root.left);
    System.out.print(root.data + " ");
    traverse(root.right);
  }

  // Function to traverse the binary tree and store its nodes' values in a set
  public static void nodeSet(Node root, Set < Integer > set) {
    // base case
    if (root == null) {
      return;
    }

    nodeSet(root.left, set);
    set.add(root.data);
    nodeSet(root.right, set);
  }

  // Function to put keys back into a set in their correct order in the BST
  public static void binaryToBST(Node root, Iterator < Integer > it) {
    if (root == null) {
      return;
    }

    binaryToBST(root.left, it);
    root.data = it.next();
    binaryToBST(root.right, it);
  }

  // Function to convert a binary tree to BST by maintaining its original structure

  public static void convertToBST(Node root) {
    // traverse the binary tree and store its nodes data in a set
    Set < Integer > set = new TreeSet < > ();
    nodeSet(root, set);

    // Iterating through the set using iterator and placing the nodes at their requires position in BST.
    Iterator < Integer > it = set.iterator();
    binaryToBST(root, it);
  }

  // Generate Tree
  public Node genTree() {
    Node root = new Node(9);
    root.left = new Node(4);
    root.right = new Node(5);
    root.left.left = new Node(15);
    root.left.right = new Node(1);
    root.right.left = new Node(3);
    root.right.right = new Node(7);

    return root;
  }

  public static void main(String[] args) {
    BinaryTreeToBST btobst = new BinaryTreeToBST();
    Node root = btobst.genTree();

    System.out.println("Given Binary Tree: ");
    traverse(root);
    System.out.print("\n\n");
    System.out.println("After converting to BST: ");
    convertToBST(root);
    traverse(root);
  }
}
